akka-actor/src/main/scala/akka/pattern/AskSupport.scala

```diff
@@ -207,53 +320,153 @@ final class AskableActorSelection(val actorSel: ActorSelection) extends AnyVal {
     case ref: InternalActorRef ⇒
-      if (timeout.duration.length <= 0)
+      if (timeout.duration.length <= 0) {
+        val message = messageFactory(ActorRef.noSender)
+        Future.failed[Any](new IllegalArgumentException(s"""Timeout length must not be negative, question not sent to [$actorRef]. Sender[$sender] sent the message of type "${message.getClass.getName}"."""))
+      } else {
+        val promiseRef = new PromiseActorRef(ref.provider)
+        val message = messageFactory(promiseRef)
+        promiseRef.setTimeout(timeout, actorRef, sender, message)
+        actorRef.tell(message, promiseRef)
+        promiseRef.future
+      }
+    case _ ⇒
+      val message = messageFactory(ActorRef.noSender)
+      Future.failed[Any](new IllegalArgumentException(s"""Unsupported recipient ActorRef type, question not sent to [$actorRef]. Sender[$sender] sent the message of type "${message.getClass.getName}"."""))
+  }
+
+  def internalSelectionAsk(actorSel: ActorSelection, messageFactory: ActorRef ⇒ Any, timeout: Timeout, sender: ActorRef): Future[Any] = actorSel.anchor match {
+    case ref: InternalActorRef ⇒
+      if (timeout.duration.length <= 0) {
+        val message = messageFactory(ActorRef.noSender)
         Future.failed[Any](
           new IllegalArgumentException(s"""Timeout length must not be negative, question not sent to [$actorSel]. Sender[$sender] sent the message of type "${message.getClass.getName}"."""))
-      else {
-        val a = PromiseActorRef(ref.provider, timeout, targetName = actorSel, message.getClass.getName, sender)
-        actorSel.tell(message, a)
-        a.result.future
+      } else {
+        val promiseRef = new PromiseActorRef(ref.provider)
+        val message = messageFactory(promiseRef)
+        promiseRef.setTimeout(timeout, actorSel, sender, message)
+        actorSel.tell(message, promiseRef)
+        promiseRef.future
       }
-    case _ ⇒ Future.failed[Any](new IllegalArgumentException(s"""Unsupported recipient ActorRef type, question not sent to [$actorSel]. Sender[$sender] sent the message of type "${message.getClass.getName}"."""))
+    case _ ⇒
+      val message = messageFactory(ActorRef.noSender)
+      Future.failed[Any](new IllegalArgumentException(s"""Unsupported recipient ActorRef type, question not sent to [$actorSel]. Sender[$sender] sent the message of type "${message.getClass.getName}"."""))
   }
 }
 
 /**
+ * INTERNAL API
+ */
+private[akka] object PromiseActorRef {
+  private case object Registering
+  private case object Stopped
+  private final case class StoppedWithPath(path: ActorPath)
+
+  private val ActorStopResult = Failure(new ActorKilledException("Stopped"))
+}
+
+/**
  * Akka private optimized representation of the temporary actor spawned to
  * receive the reply to an "ask" operation.
  *
  * INTERNAL API
  */
-private[akka] final class PromiseActorRef private (val provider: ActorRefProvider, val result: Promise[Any], _mcn: String)
```

Comment:
I think this class will need to remain binary compatible (it is used by prominent “naughty” projects).